### PR TITLE
Script to run all examples.

### PR DESCRIPTION
**Pull request by cqc-alec**
The script is run nightly on the CI (excluding notebooks that require IBMQ credentials), and can also be run locally to test all notebooks.

See the addition to the `README` for an explanation.

Note that the primary sources are now Python scripts (with comments in Markdown) and the notebook files are generated from these.

(Reviewer: no need to review the ipynb files, since correct generation of these is checked by the `check-examples` script. The very first commit in this PR was the generation of the Python files directly from the existing notebooks; the rest are fixing up formatting and making the scripts for the CI checks. There are no substantive changes to the notebooks except for the replacement of IBMQ jobs with IBMQ emulator jobs towards the end of `backends_example`.)

### CODE REVIEW

File: examples/Forest_portability_example.ipynb [hunk 1]
```diff
@@ -1,268 +1 @@
...
... "metadata": {}, "outputs": [], "source": ["tk_circ.measure_all()"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["from pytket.extensions.pyquil import ForestBackend"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["aspen_backend = ForestBackend(\"9q-square\", simulator=True)\n", "aspen_backend.compile_circuit(tk_circ)"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["counts = aspen_backend.get_counts(tk_circ, 2000)\n", "print(counts)"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Note that attempting to connect to a live quantum device (using a `ForestBackend` with `as_simulator=False`) will fail unless it is running from a QMI instance during a reservation for the named lattice."]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.4"}}, "nbformat": 4, "nbformat_minor": 2}
```
Comment (cqc-melf) [inline examples/Forest_portability_example.ipynb:1]: I think here is a problem with the format of this file?

File: examples/ansatz_sequence_example.ipynb [hunk 2]
```diff
@@ -1,318 +1 @@
...
...uit\n", ")\n", "print(last_circuit.get_commands())"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["print(\"Last CX Depth: {}\".format(last_circuit.depth_by_type(OpType.CX)))\n", "print(\"Last CX Count: {}\".format(last_circuit.n_gates_of_type(OpType.CX)))"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Other than some single-qubit Cliffords we acquired via synthesis, you can check that this gives us the same circuit structure as our `Transform.DecomposeBoxes` method! It is a suboptimal synthesis method.<br>\n", "<br>\n", "As with the `gen_term_sequence` method, we recommend playing around with the arguments and seeing what circuits come out. Typically we find that `PauliSynthStrat.Sets` and `CXConfigType.Tree` work the best, although routing can affect this somewhat."]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.4"}}, "nbformat": 4, "nbformat_minor": 2}
```
Comment (cqc-melf) [inline examples/ansatz_sequence_example.ipynb:1]: See above...

File: examples/backends_example.ipynb [hunk 3]
```diff
@@ -1,712 +1 @@
...
...ch loop of a variational algorithm), or removing individual results with `Backend.pop_result`."]}, {"cell_type": "markdown", "metadata": {}, "source": ["The backends in `pytket` are designed to be as similar to one another as possible. The example above using physical devices can be run entirely on a simulator by swapping out the `IBMQBackend` constructor for any other backend supporting shot outputs (e.g. `AerBackend`, `ProjectQBackend`, `ForestBackend`), or passing it the name of a different device. Furthermore, using pytket it is simple to convert between handling shot tables, counts maps and statevectors.<br>\n", "<br>\n", "For more information on backends and other `pytket` features, read our [documentation](https://cqcl.github.io/pytket) or see the other examples on our [GitHub repo](https://github.com/CQCL/pytket)."]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.4"}}, "nbformat": 4, "nbformat_minor": 2}
```
Comment (cqc-melf) [inline examples/backends_example.ipynb:1]: See above

File: examples/circuit_analysis_example.ipynb [hunk 4]
```diff
@@ -1,434 +1 @@
...
...e": ["From the `Op` we can read off the string representation (in normal or Latex form), the parameters and the type:"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["op0.get_name()  # normal form"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["op0.get_name(latex=True)  # Latex form"]}, {"cell_type": "code", "execution_count": null, "metadata": {}, "outputs": [], "source": ["op0.type, op0.params"]}, {"cell_type": "markdown", "metadata": {}, "source": ["Note that some compilation passes introduce implicit wire swaps at the end of the circuit, which are not represented in the command list. (The internal representation of the circuit as a directed acyclic graph reduces explicit permutations of qubits to implicit features of the graph.)"]}], "metadata": {"kernelspec": {"display_name": "Python 3", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.6.4"}}, "nbformat": 4, "nbformat_minor": 2}
```
Comment (cqc-melf) [inline examples/circuit_analysis_example.ipynb:1]: See above

File: examples/README.md [hunk 5]
```diff
@@ -58,4 +58,22 @@ Using tomography to analyse the effect of noise channels when iterating the Enta
...
+from the Python scripts; and then attempts to run all the scripts. For those
+scripts that require credentials, it first tries to load them, and skips the
+test if it fails. For those that require containers to be running, it first
+tries to start them, and skips the test if it fails.
+
+To add a new script, you can either create the Python file and convert it using
+`gen-nb` or create a notebook file, convert it to Python using `p2j -r`, and
+then run `gen-nb` to convert back. In either case, please check that the
+formatting in the notebook is satisfactory. Then add the name of the new
+notebook to `maintained-notebooks.txt`.
```
Comment (cqc-melf) [inline examples/README.md:79]: You should mention that the name of the notebook needs to be added to the workflow in line 13-16 or 28 if login credentials are needed.

File: examples/Forest_portability_example.ipynb (hunk 1 above)
Comment (cqc-alec) [reply examples/Forest_portability_example.ipynb:1]: What is the problem? It displays OK for me in a browser.
(Note these files are autogenerated, we don't control the formatting.)

Comment (cqc-melf) [reply examples/Forest_portability_example.ipynb:1]: I thought it is maybe a bit hard to read without line breaks. Maybe black could split that up after the generation? If that is not possible just leave it as it is.

Comment (cqc-alec) [reply examples/Forest_portability_example.ipynb:1]: Well these files aren't meant to be read by us: it is the Python files (in the `python` directory) that we will read and edit.
That was one reason to make this change: ipynb files are horrible to edit and diff at the best of times.

Comment (cqc-melf) [reply examples/Forest_portability_example.ipynb:1]: Okay, thank you for the explanations. I am now fine with this.

**Review by cqc-melf — APPROVED** [review]

**Pull request merged**